File: backend/src/main/java/br/ufpr/sept/so2/modules/presenca/application/ConfirmarPresencaUseCase.java
package br.ufpr.sept.so2.modules.presenca.application;

import br.ufpr.sept.so2.modules.iam.application.ports.OutboxPort;
import br.ufpr.sept.so2.modules.iam.application.ports.PasswordHasher;
import br.ufpr.sept.so2.modules.presenca.application.ports.EventoRepository;
import br.ufpr.sept.so2.modules.presenca.application.ports.PresencaRepository;
import br.ufpr.sept.so2.modules.presenca.domain.AttendanceMode;
import br.ufpr.sept.so2.modules.presenca.domain.Evento;
import br.ufpr.sept.so2.modules.presenca.domain.FasePresenca;
import br.ufpr.sept.so2.modules.presenca.domain.Presenca;
import br.ufpr.sept.so2.shared.domain.exception.ConflitoEstadoException;
import br.ufpr.sept.so2.shared.domain.exception.RecursoNaoEncontradoException;
import br.ufpr.sept.so2.shared.infrastructure.Uuids;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.OffsetDateTime;
import java.util.List;
import java.util.Map;
import java.util.UUID;

@Service
public class ConfirmarPresencaUseCase {

    private final EventoRepository eventoRepository;
    private final PresencaRepository presencaRepository;
    private final PasswordHasher passwordHasher;
    private final OutboxPort outboxPort;
    private final ObjectMapper objectMapper;

    public ConfirmarPresencaUseCase(
            EventoRepository eventoRepository,
            PresencaRepository presencaRepository,
            PasswordHasher passwordHasher,
            OutboxPort outboxPort,
            ObjectMapper objectMapper
    ) {
        this.eventoRepository = eventoRepository;
        this.presencaRepository = presencaRepository;
        this.passwordHasher = passwordHasher;
        this.outboxPort = outboxPort;
        this.objectMapper = objectMapper;
    }

    @Transactional
    public ObterSessaoPresencaUseCase.SessaoPresenca execute(
            UUID eventoId,
            UUID usuarioId,
            String pin,
            String deviceUuid,
            String faseRaw
    ) {
        Evento evento = eventoRepository.findById(eventoId)
                .orElseThrow(() -> new RecursoNaoEncontradoException("Evento não encontrado."));
        FasePresenca fase = FasePresenca.from(faseRaw);
        if (presencaRepository.existsByEventoUsuarioFase(eventoId, usuarioId, fase)) {
            throw new ConflitoEstadoException("A presença desta fase já foi confirmada.");
        }
        if (presencaRepository.existsByEventoAndDeviceDeOutroUsuario(eventoId, deviceUuid, usuarioId)) {
            throw new ConflitoEstadoException("Este dispositivo já foi usado neste evento.");
        }
        boolean pinValido = passwordHasher.matches(pin, evento.getPinHash());
        if (!pinValido) {
            passwordHasher.matchesDummy();
        }
        evento.garantirConfirmacao(AttendanceMode.SECRET_SINGLE, fase, OffsetDateTime.now(), pinValido);
        Presenca presenca = Presenca.registrar(
                Uuids.v7(),
                eventoId,
                usuarioId,
                fase,
                deviceUuid,
                OffsetDateTime.now()
        );
        presencaRepository.save(presenca);
        outboxPort.enqueue("presenca.confirmada", payload(eventoId, usuarioId, fase, presenca.getId()));
        List<FasePresenca> fases = presencaRepository.findByEventoAndUsuario(eventoId, usuarioId).stream()
                .map(Presenca::getFase)
                .toList();
        return new ObterSessaoPresencaUseCase.SessaoPresenca(evento, fases);
    }

    private String payload(UUID eventoId, UUID usuarioId, FasePresenca fase, UUID presencaId) {
        try {
            return objectMapper.writeValueAsString(Map.of(
                    "eventoId", eventoId.toString(),
                    "usuarioId", usuarioId.toString(),
                    "fase", fase.name(),
                    "presencaId", presencaId.toString()
            ));
        } catch (JsonProcessingException ex) {
            return "{\"eventoId\":\"" + eventoId + "\"}";
        }
    }
}

File: backend/src/main/kotlin/br/ufpr/sept/so2/modules/presenca/application/ConfirmarPresencaUseCase.kt
package br.ufpr.sept.so2.modules.presenca.application

import br.ufpr.sept.so2.modules.formativas.application.ports.FormativaPorPresencaPort
import br.ufpr.sept.so2.modules.iam.application.ports.AuditLogPort
import br.ufpr.sept.so2.modules.iam.application.ports.OutboxPort
import br.ufpr.sept.so2.modules.iam.application.ports.PasswordHasher
import br.ufpr.sept.so2.modules.presenca.application.ports.EventoRepository
import br.ufpr.sept.so2.modules.presenca.application.ports.PresencaRepository
import br.ufpr.sept.so2.modules.presenca.domain.Evento
import br.ufpr.sept.so2.modules.presenca.domain.FasePresenca
import br.ufpr.sept.so2.modules.presenca.domain.Presenca
import br.ufpr.sept.so2.shared.domain.exception.ConflitoEstadoException
import br.ufpr.sept.so2.shared.domain.exception.RecursoNaoEncontradoException
import br.ufpr.sept.so2.shared.infrastructure.Uuids
import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.OffsetDateTime
import java.util.UUID

@Service
class ConfirmarPresencaUseCase(
    private val eventoRepository: EventoRepository,
    private val presencaRepository: PresencaRepository,
    private val passwordHasher: PasswordHasher,
    private val outboxPort: OutboxPort,
    private val auditLogPort: AuditLogPort,
    private val objectMapper: ObjectMapper,
    private val formativaPorPresencaPort: FormativaPorPresencaPort,
) {
    @Transactional
    fun execute(
        eventoId: UUID,
        usuarioId: UUID,
        pin: String?,
        token: String?,
        deviceUuid: String,
        faseRaw: String,
        ip: String?,
    ): ObterSessaoPresencaUseCase.SessaoPresenca {
        val evento = eventoRepository.findById(eventoId)
            .orElseThrow { RecursoNaoEncontradoException("Evento não encontrado.") }
        val fase = FasePresenca.from(faseRaw)
        if (presencaRepository.existsByEventoUsuarioFase(eventoId, usuarioId, fase)) {
            throw ConflitoEstadoException("A presença desta fase já foi confirmada.")
        }
        if (presencaRepository.existsByEventoAndDeviceDeOutroUsuario(eventoId, deviceUuid, usuarioId)) {
            throw ConflitoEstadoException("Este dispositivo já foi usado neste evento.")
        }
        if (fase == FasePresenca.SAIDA &&
            !presencaRepository.existsByEventoUsuarioFase(eventoId, usuarioId, FasePresenca.ENTRADA)
        ) {
            throw ConflitoEstadoException("Confirmação de saída exige entrada registrada.")
        }
        val informado = if (evento.attendanceMode.isSecret()) pin else token
        val segredoValido = passwordHasher.matches(informado, evento.pinHash)
        if (!segredoValido) {
            passwordHasher.matchesDummy()
        }
        val agora = OffsetDateTime.now()
        evento.garantirConfirmacao(fase, agora, segredoValido)
        val presenca = Presenca.registrar(
            Uuids.v7(),
            eventoId,
            usuarioId,
            fase,
            deviceUuid,
            agora,
        )
        presencaRepository.save(presenca)
        val payload = payload(eventoId, usuarioId, fase, presenca.id)
        outboxPort.enqueue("presenca.confirmada", payload)
        auditLogPort.append("presenca.confirmada", usuarioId, payload, ip)
        if (presencaCompleta(evento, fase)) {
            formativaPorPresencaPort.criarPendenteSeAusente(
                eventoId,
                usuarioId,
                evento.titulo,
                evento.cargaHoraria,
            )
        }
        val fases = presencaRepository.findByEventoAndUsuario(eventoId, usuarioId).map { it.fase }
        return ObterSessaoPresencaUseCase.SessaoPresenca(evento, fases)
    }

    private fun payload(eventoId: UUID, usuarioId: UUID, fase: FasePresenca, presencaId: UUID): String =
        try {
            objectMapper.writeValueAsString(
                mapOf(
                    "eventoId" to eventoId.toString(),
                    "usuarioId" to usuarioId.toString(),
                    "fase" to fase.name,
                    "presencaId" to presencaId.toString(),
                ),
            )
        } catch (_: JsonProcessingException) {
            "{\"eventoId\":\"$eventoId\"}"
        }

    companion object {
        fun presencaCompleta(evento: Evento, fase: FasePresenca): Boolean =
            if (evento.attendanceMode.isDual()) {
                fase == FasePresenca.SAIDA
            } else {
                fase == FasePresenca.ENTRADA
            }
    }
}
